components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs

```diff
@@ -0,0 +1,671 @@
+        /// <param name="throwOnError">if set to <c>true</c> a exception will be thrown on an error.</param>
+        /// <returns>
+        /// The references found. Null if an error occurred.
+        /// </returns>
+        public static ReferenceDescriptionCollection Browse(Session session, BrowseDescription nodeToBrowse, bool throwOnError) {
+            return Browse(session, null, nodeToBrowse, throwOnError);
+        }
+
+        /// <summary>
+        /// Browses the address space and returns the references found.
+        /// </summary>
+        public static ReferenceDescriptionCollection Browse(Session session, ViewDescription view, BrowseDescription nodeToBrowse, bool throwOnError) {
+            try {
+                ReferenceDescriptionCollection references = new ReferenceDescriptionCollection();
+
+                // construct browse request.
+                BrowseDescriptionCollection nodesToBrowse = new BrowseDescriptionCollection();
+                nodesToBrowse.Add(nodeToBrowse);
+
+                // start the browse operation.
+                BrowseResultCollection results = null;
+                DiagnosticInfoCollection diagnosticInfos = null;
+
+                session.Browse(
+                    null,
+                    view,
+                    0,
+                    nodesToBrowse,
+                    out results,
+                    out diagnosticInfos);
+
+                ClientBase.ValidateResponse(results, nodesToBrowse);
+                ClientBase.ValidateDiagnosticInfos(diagnosticInfos, nodesToBrowse);
+
+                do {
+                    // check for error.
+                    if (StatusCode.IsBad(results[0].StatusCode)) {
+                        throw new ServiceResultException(results[0].StatusCode);
+                    }
+
+                    // process results.
+                    for (int ii = 0; ii < results[0].References.Count; ii++) {
+                        references.Add(results[0].References[ii]);
+                    }
+
+                    // check if all references have been fetched.
+                    if (results[0].References.Count == 0 || results[0].ContinuationPoint == null) {
+                        break;
+                    }
+
+                    // continue browse operation.
+                    ByteStringCollection continuationPoints = new ByteStringCollection();
+                    continuationPoints.Add(results[0].ContinuationPoint);
+
+                    session.BrowseNext(
+                        null,
+                        false,
+                        continuationPoints,
+                        out results,
+                        out diagnosticInfos);
+
+                    ClientBase.ValidateResponse(results, continuationPoints);
+                    ClientBase.ValidateDiagnosticInfos(diagnosticInfos, continuationPoints);
+                }
+                while (true);
```

Comment:
We should have an timeout to not be forever in this loop

Reply:
Fixed